extensions/amp-analytics/0.1/linker-manager.js

```diff
@@ -0,0 +1,218 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ExpansionOptions, variableServiceFor} from './variables';
+import {Priority} from '../../../src/service/navigation';
+import {Services} from '../../../src/services';
+import {addParamToUrl} from '../../../src/url';
+import {createLinker} from './linker';
+import {dict} from '../../../src/utils/object';
+import {user} from '../../../src/log';
+
+const TAG = 'amp-analytics-linker';
+
+/**
+ * The name of the Google CID API as it appears in the meta tag to opt-in.
+ * @const @private {string}
+ */
+const GOOGLE_CID_API_META_NAME = 'amp-google-client-id-api';
+
+export class LinkerManager {
+
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {JsonObject} config
+   * @param {?string} type
+   */
+  constructor(ampdoc, config, type) {
+    /** @private */
+    this.ampdoc_ = ampdoc;
+
+    /** @private {JsonObject} */
+    this.config_ = config;
+
+    /** @private {?string} */
+    this.type_ = type;
+
+    /** @private {!Array<Promise>} @visibleForTesting */
+    this.allLinkerPromises_ = [];
+
+    /** @private {JsonObject} */
+    this.resolvedLinkers_ = dict();
+  }
+
+
+  /**
+   * Start resolving any macros that may exist in the linker configuration
+   * and register the callback with the navigation service. Since macro
+   * resolution is aynchronous the callback may be looking for these values
+   * before they are ready.
+   */
+  init() {
+    if (!this.config_['linkers']) {
+      return;
+    }
+
+    const linkerNames = Object.keys(this.config_['linkers']);
+
+    if (!linkerNames.length) {
+      return user().error(TAG,
```

Comment:
It might worth to support syntax below to completely disable linkers:

```
{
  linkers: false
}
```

So, we might not want to print this error.

Reply:
Should we check both? `linkers: false` is good config, but `linkers: {}` is bad?

Reply:
Actually I think this is covered by line 65. `if (!this.config_['linkers'])` should cover the `linkers: false` case and exit before this code

Reply:
honestly i don't see strong reason to now allow `linkers: {}`, especially with increasing binary size.

Reply:
ok removed.